flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/nn/QuadTree.scala

```diff
@@ -0,0 +1,352 @@
+      }
+    } else {
+      insertRecur(queryPoint, node.children(node.whichChild(queryPoint)))
+    }
+  }
+
+  /**
+   * Used to zoom in on a region near a test point for a fast KNN query.
+   * This capability is used in the KNN query to find k "near" neighbors n_1,...,n_k, from
+   * which one computes the max distance D_s to queryPoint.  D_s is then used during the
+   * kNN query to find all points within a radius D_s of queryPoint using searchNeighbors.
+   * To find the "near" neighbors, a min-heap is defined on the leaf nodes of the leaf
+   * nodes of the minimal bounding box of the queryPoint. The priority of a leaf node
+   * is an appropriate notion of the distance between the test point and the node,
+   * which is defined by minDist(queryPoint),
+   *
+   * @param queryPoint a test point for which the method finds the minimal bounding
+   *                   box that queryPoint lies in and returns elements in that boxes
+   *                   siblings' leaf nodes
+   * @return
+   */
+  def searchNeighborsSiblingQueue(queryPoint: Vector): ListBuffer[Vector] = {
+    val ret = new ListBuffer[Vector]
+    // edge case when the main box has not been partitioned at all
+    if (root.children == null) {
+      root.nodeElements.clone()
+    } else {
+      val nodeQueue = new PriorityQueue[(Double, Node)]()(Ordering.by(x => x._1))
+      searchRecurSiblingQueue(queryPoint, root, nodeQueue)
+
+      var count = 0
+      while (count < maxPerBox) {
+        val dq = nodeQueue.dequeue()
+        if (dq._2.nodeElements.nonEmpty) {
+          ret ++= dq._2.nodeElements
+          count += dq._2.nodeElements.length
+        }
+      }
+      ret
+    }
+  }
+
+  /**
+   *
+   * @param queryPoint point under consideration
+   * @param node node that queryPoint lies in
+   * @param nodeQueue defined in searchSiblingQueue, this stores nodes based on their
+   *                  distance to node as defined by minDist
+   */
+  private def searchRecurSiblingQueue(
+    queryPoint: Vector,
+    node: Node,
+    nodeQueue: PriorityQueue[(Double, Node)]) {
+    if (node.children != null) {
+      for (child <- node.children; if child.contains(queryPoint)) {
+        if (child.children == null) {
+          for (c <- node.children) {
+            minNodes(queryPoint, c, nodeQueue)
+          }
+        } else {
+          searchRecurSiblingQueue(queryPoint, child, nodeQueue)
+        }
+      }
+    }
+  }
+
+  /**
+   * Goes down to minimal bounding box of queryPoint, and add elements to nodeQueue
+   *
+   * @param queryPoint point under consideration
+   * @param node node that queryPoint lies in
+   * @param nodeQueue PriorityQueue that stores all points in minimal bounding box of queryPoint
+   */
+  private def minNodes(
+    queryPoint: Vector,
+    node: Node,
+    nodeQueue: PriorityQueue[(Double, Node)]) {
+    if (node.children == null) {
+      nodeQueue += ((-node.minDist(queryPoint), node))
```

Comment:
This is to make a min-heap.  I take all the siblings' leaf nodes of the node that `queryPoint` lies in, and of those nodes I want to extract the ones closest to `queryPoint`.